draft-ietf-ppm-dap.md

```diff
@@ -1186,27 +1271,157 @@ the last round is reached and each Aggregator computes an output share. All
+inputs and/or a prep step sent by the peer. The initial state is `START`.
+
+~~~ transition
+State START
+
+Input
+
+  vdaf_verify_key: VDAF verification key associated with the task
+  is_leader: boolean indicating whether the host is the Leader
+  agg_param: aggregation parameter for the aggregation job
+  report_id: report ID
+  public_share: report public share
+  host_input_share: host's input share for the report
+
+def VdafPrepInit(vdaf_verify_key,
+                 is_leader,
+                 agg_param,
+                 report_id,
+                 public_share,
+                 host_input_share):
+  try:
+    host_prep_state = Vdaf.prep_init(
+      vdaf_verify_key,
+      0 if is_leader else 1,
+      agg_param,
+      report_id,
+      public_share,
+      host_input_share,
+    )
+    state CONTINUED(0, host_prep_state)
+  except:
+    state REJECTED
+~~~
+
+If the state is `REJECTED`, then processing halts. Otherwise, if the state is
+`CONTINUE`, then processing continues using the following transition rule until
+a terminal state is reached.
+
+~~~ transition
+State CONTINUED(round, host_prep_state)
+
+Input
+
+  is_leader: boolean indicating whether the host is the Leader
+  agg_param: aggregation parameter for the aggregation job
+  inbound: PrepareStep (or None)
+
+Output
+
+  outbound: PrepareStep (or None)
+
+def VdafPrepNext(is_leader, agg_param, inbound):
+  if type(inbound) == PrepareStep.initialize:
+    prep_msg, peer_prep_share = None, inbound.prep_share
+  elif type(inbound) = PrepareStep.continue:
+    prep_msg, peer_prep_share = inbound.prep_msg, inbound.prep_share
+  elif type(inbound) = PrepareStep.finish:
+    prep_msg, peer_prep_share = inbound.prep_msg, None
+  else:
+    prep_msg, peer_prep_share = None, None
+
+  try:
+    /* If prep_msg == None then this is the start of the
+     * first round. Otherwise, this is the prep message
+     * from the end of the previous round. */
+    out = Vdaf.prep_next(host_prep_state, prep_msg)
+    if round == Vdaf.ROUNDS:
+      state FINISHED(out)
+      /* The Leader reaches this state upon processing a response
+       * from the Helper, so there is no outbound message. */
+      return None if is_leader else PrepareStep.finished
+    (host_prep_state, host_prep_share) = out
+    round += 1
+
+    if peer_prep_share != None:
+      prep_shares = [peer_prep_share, host_prep_share]
+      if is_leader:
+        prep_shares.reverse()
+      prep_msg = Vdaf.prep_shares_to_prep(
+        agg_param,
+        prep_shares,
+      )
+      out = Vdaf.prep_next(host_prep_state, prep_msg)
+      if round == Vdaf.ROUNDS:
+        state FINISHED(out)
+        return PrepareStep.finish(prep_msg)
+      (host_prep_state, host_prep_share) = out
+
+    state CONTINUED(round, host_prep_state)
+    if is_leader and round == 1:
+      return PrepareStep.initialize(host_prep_share)
+    else:
+      return PrepareStep.continue(prep_msg, host_prep_share)
```

Comment:
As noted in https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/451#issuecomment-1531741943, writing things this way effectively requires the Leader to store both the `prep_msg` and the `prep_share` to durable storage, barring implementors effectively splitting this function up to run part of it in one DAP round and part of it in another (which IMO may be a stretch to expect of implementors). #393 is written in a way that stores only `prep_msg`.

Maybe we can include an implementation note that the final `prep_next` call can actually be made after storing `prep_msg` to storage, for implementations that "checkpoint" between rounds?

Reply:
> As noted in [#451 (comment)](https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/issues/451#issuecomment-1531741943), writing things this way effectively requires the Leader to store both the `prep_msg` and the `prep_share` to durable storage, barring implementors effectively splitting this function up to run part of it in one DAP round and part of it in another (which IMO may be a stretch to expect of implementors). #393 is written in a way that stores only `prep_msg`.

I don't quite understand what you mean by durable storage. The intent of these methods is to describe the semantics of 393, i.e., I don't expect the protocol flow to be different.

Reply:
I think what is returned by `DapPrepNext` is effectively what a leader stores for retries -- `DapPrepNext` is the last step for leader initialization & continuation, and the result of that call is what is sent to the helper at the beginning of the next round of continuation.

But an implementation could effectively "pause" computation of `DapPrepNext` before the final `VDAF.prep_next` call, and only need to keep around the `prep_msg`, recomputing the `host_prep_share` if/when a retry occurs. This would help implementations that checkpoint each round to durable storage to e.g. survive unexpected process failures.

This is an implementation choice, so I don't think we need to make a change, but an implementation note would be helpful as I think otherwise it would be challenging for implementors to realize this optimization is possible.

Reply:
(As discussed offline) The plan here will be to first try to make this point clear in the normative protocol text. Let's keep this section focused on the state machine.